TumbleBitSetup/PoupardStern.cs

```diff
@@ -0,0 +1,263 @@
+            BigInteger qSub1 = q.Subtract(BigInteger.One);
+            BigInteger phi = pSub1.Multiply(qSub1);
+
+            // Check that N>2^{|N|-1}
+            if (!(Modulus.CompareTo(lowerLimit) > 0))
+                throw new ArgumentOutOfRangeException("Bad RSA modulus N");
+
+            // (N-phi)*2^k << N
+            if (!(Modulus.Subtract(phi).Multiply(Two.Pow(k)).CompareTo(Modulus) < 0))
+                throw new ArgumentOutOfRangeException("Bad RSA modulus N");
+
+            // Generate K
+            GetK(k, out int BigK);
+
+            // Initialize list of x and z values
+            BigInteger[] xValues = new BigInteger[BigK];
+            BigInteger[] zValues = new BigInteger[BigK];
+
+            // Generate the list of z Values
+            for (int i = 0; i < BigK; i++)
+                zValues[i] =  SampleFromZnStar(pubKey, psBytes, i, k, ModulusBitLength);
+
+            for (;;)
+            {
+                // Generate r
+                GetR(ModulusBitLength, out BigInteger r);
+
+                for (int i = 0; i < BigK; i++)
+                    // Compute x_i
+                    xValues[i] = zValues[i].ModPow(r, Modulus);
+
+                // Compute w
+                GetW(pubKey, psBytes, xValues, k, ModulusBitLength, out BigInteger w);
+
+                // Compute y
+                // Make sure the n == N in step 5, page 14 (a typo probably)
+                y = r.Add(Modulus.Subtract(phi)).Multiply(w);
+
+                // if y >= 2^{ |N| - 1 }
+                if (y.CompareTo(lowerLimit) >= 0)
+                    continue;
+                
+                // if y < 0
+                if (y.CompareTo(BigInteger.Zero) < 0)
+                    continue;
+
+                return new Tuple<BigInteger[], BigInteger>(xValues, y);
+            }
+            
+        }
+
+        /// <summary>
+        /// Verifying Algorithm specified in page 13 (3.3) of the setup
+        /// </summary>
+        /// <param name="pubKey">Public key used</param>
+        /// <param name="xValues">List of x_i values</param>
+        /// <param name="y">The value y as specified in the setup</param>
+        /// <param name="keyLength">The size of the RSA key in bits</param>
+        /// <param name="ps">public string specified in the setup</param>
+        /// <param name="k">Security parameter specified in the setup.</param>
+        /// <returns>true if the xValues verify, false otherwise</returns>
+        public static bool Verifying(RsaPubKey pubKey, BigInteger[] xValues, BigInteger y, int keyLength, byte[] psBytes, int k = 128)
+        {
+            BigInteger rPrime;
+            BigInteger lowerLimit = BigInteger.Two.Pow(keyLength - 1);
+            var Modulus = pubKey._pubKey.Modulus;
+            var Exponent = pubKey._pubKey.Exponent;
+
+            // Checking that:
+            // if y < 2^{ |N| - 1 }
+            if (!(y.CompareTo(lowerLimit) < 0))  
+                return false;
```

Comment:
it's "lowerLimit" because it's used to compare with N in the function ([here](https://github.com/osagga/TumbleBitSetup/blob/PoupardSternProtocol/TumbleBitSetup/PoupardStern.cs#L118)). So it's a lowerLimit for N and an upperLimit for y, so I'm not sure what should I name this value :)